File: frontend/src/types/quotation.ts
export interface PriceDetail {
  amount: number;
  currency: string;
  symbol?: string;
}

export interface QuoteItem {
  line_number: number;
  part_number: string;
  description: string;
  quantity?: number | null;
  unit?: string | null;
  unit_price?: number | null;
  unit_price_detail?: PriceDetail | null;
  currency?: string;
  currency_symbol?: string;
  discount_percent?: number | null;
  total_price?: number | null;
  total_price_detail?: PriceDetail | null;
  commodity_code?: string | null;
  equipment_group?: string | null;
  item_type?: 'part' | 'charge' | 'service' | 'non_inventory' | string;
  source_page?: number;
  confidence?: Record<string, number>;
  status: 'verified' | 'warning' | 'error';
  validation_notes?: string[];
}

export interface EquipmentGroup {
  name: string;
  serial_numbers: string[];
  line_numbers: number[];
  line_start?: number;
  line_end?: number;
}

export interface QuotationData {
  quote_number?: string;
  quote_date?: string;
  expiry_date?: string;
  supplier_name?: string;
  customer?: string;
  payment_terms?: string;
  delivery_terms?: string;
  sales_person?: string;
  handled_by?: string;
  email?: string;
  currency?: string;
  currency_symbol?: string;
  items: QuoteItem[];
  equipment_groups: EquipmentGroup[];
  packaging_cost?: number | null;
  miscellaneous_charges?: number | null;
  lines_total?: number | null;
  grand_total?: number | null;
  pdf_item_count?: number;
  extracted_item_count?: number;
  raw_pdf_text?: string;
  source_pdf_id?: string;
  template_id?: string;
}

export interface ValidationIssue {
  item_line?: number;
  field: string;
  issue_type: 'error' | 'warning' | 'info';
  message: string;
  expected_value?: any;
  actual_value?: any;
}

export interface ValidationSummary {
  is_valid: boolean;
  overall_confidence: number;
  errors_count: number;
  warnings_count: number;
  pdf_item_count: number;
  extracted_item_count: number;
  items_ready_for_excel: number;
  currency: string;
  currency_symbol: string;
  price_validation_passed: number;
  part_number_validation_passed: number;
  total_items_validated: number;
  issues: ValidationIssue[];
}

export interface ExtractionResponse {
  success: boolean;
  quotation: QuotationData;
  validation: ValidationSummary;
  processing_time_seconds: number;
  extracted_page_count: number;
  extraction_source: string;
}

export interface ColumnMappingConfig {
  line_number_column?: string;
  part_number_column?: string;
  description_column?: string;
  quantity_column?: string;
  unit_column?: string;
  unit_price_column?: string;
  discount_column?: string;
  total_price_column?: string;
  currency_column?: string;
  start_row: number;
  quote_number_cell?: string;
  customer_cell?: string;
  date_cell?: string;
  currency_cell?: string;
}

export interface TemplateAnalysisResponse {
  template_id: string;
  sheet_names: string[];
  detected_headers: Record<string, string>;
  suggested_mapping: ColumnMappingConfig;
  preview_rows: any[][];
}

export interface GenerateExcelRequest {
  quotation: QuotationData;
  template_id: string;
  pricing_mode: 'quoted_price' | 'supplier_discount' | 'add_margin';
  margin_percent: number;
  supplier_discount_percent?: number;
  custom_mapping?: ColumnMappingConfig;
}

export interface GenerateExcelResponse {
  success: boolean;
  download_url: string;
  filename: string;
  file_id: string;
  total_items_written: number;
  grand_total_written: number;
}

export interface ConversionRecord {
  id: string;
  quote_number: string;
  customer: string;
  source_pdf_filename: string;
  source_pdf_id: string;
  template_id: string;
  quote_date: string;
  created_at: string;
  updated_at: string;
  status: 'generated' | 'verified' | 'warning' | 'error' | 'draft';
  items_count: number;
  currency: string;
  grand_total: number;
  pricing_mode?: string;
  excel_file_id?: string;
  excel_filename?: string;
  quotation_data?: QuotationData;
  validation_summary?: ValidationSummary;
}

export interface HistoryResponse {
  success: boolean;
  records: ConversionRecord[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

File: backend/app/schemas/quotation.py
from typing import Optional, Any, Union
from pydantic import BaseModel, Field, field_validator


class PriceDetail(BaseModel):
    amount: float
    currency: str = "EUR"
    symbol: Optional[str] = "€"


class FieldConfidence(BaseModel):
    field: str
    value: Any
    confidence: float = 1.0
    status: str = "verified"  # "verified", "warning", "error"
    message: Optional[str] = None


class QuoteItem(BaseModel):
    line_number: int
    part_number: str = Field(..., description="Part number as exact string. Must preserve leading zeros.")
    description: str = ""
    quantity: Optional[float] = None
    unit: Optional[str] = None
    unit_price: Optional[float] = None
    unit_price_detail: Optional[PriceDetail] = None
    currency: Optional[str] = "EUR"
    currency_symbol: Optional[str] = "€"
    discount_percent: Optional[float] = None
    total_price: Optional[float] = None
    total_price_detail: Optional[PriceDetail] = None
    commodity_code: Optional[str] = None
    equipment_group: Optional[str] = None
    item_type: Optional[str] = "part"  # "part", "charge", "service", "non_inventory"
    source_page: Optional[int] = 1
    confidence: dict[str, float] = Field(default_factory=dict)
    status: str = "verified"  # "verified", "warning", "error"
    validation_notes: list[str] = Field(default_factory=list)

    @field_validator("part_number", mode="before")
    @classmethod
    def enforce_string_part_number(cls, v: Any) -> str:
        if v is None:
            return ""
        return str(v).strip()

    @field_validator("commodity_code", mode="before")
    @classmethod
    def enforce_string_commodity_code(cls, v: Any) -> Optional[str]:
        if v is None or str(v).strip() == "" or str(v).strip().lower() in ("none", "null"):
            return None
        return str(v).strip()

    @field_validator("unit_price", mode="before")
    @classmethod
    def parse_unit_price(cls, v: Any) -> Optional[float]:
        if isinstance(v, dict):
            return float(v.get("amount", 0.0)) if v.get("amount") is not None else None
        if v is None or str(v).strip() == "" or str(v).strip().lower() in ("none", "null"):
            return None
        try:
            return float(v)
        except (ValueError, TypeError):
            return None

    @field_validator("total_price", mode="before")
    @classmethod
    def parse_total_price(cls, v: Any) -> Optional[float]:
        if isinstance(v, dict):
            return float(v.get("amount", 0.0)) if v.get("amount") is not None else None
        if v is None or str(v).strip() == "" or str(v).strip().lower() in ("none", "null"):
            return None
        try:
            return float(v)
        except (ValueError, TypeError):
            return None


class EquipmentGroup(BaseModel):
    name: str
    serial_numbers: list[str] = Field(default_factory=list)
    line_numbers: list[int] = Field(default_factory=list)
    line_start: Optional[int] = None
    line_end: Optional[int] = None


class QuotationData(BaseModel):
    quote_number: Optional[str] = None
    quote_date: Optional[str] = None
    expiry_date: Optional[str] = None
    supplier_name: Optional[str] = None
    customer: Optional[str] = None
    payment_terms: Optional[str] = None
    delivery_terms: Optional[str] = None
    sales_person: Optional[str] = None
    handled_by: Optional[str] = None
    email: Optional[str] = None
    currency: Optional[str] = "EUR"
    currency_symbol: Optional[str] = "€"
    items: list[QuoteItem] = Field(default_factory=list)
    equipment_groups: list[EquipmentGroup] = Field(default_factory=list)
    packaging_cost: Optional[float] = None
    miscellaneous_charges: Optional[float] = None
    lines_total: Optional[float] = None
    grand_total: Optional[float] = None
    pdf_item_count: Optional[int] = None
    extracted_item_count: Optional[int] = None
    raw_pdf_text: Optional[str] = None
    source_pdf_id: Optional[str] = None
    template_id: Optional[str] = None


class ValidationIssue(BaseModel):
    item_line: Optional[int] = None
    field: str
    issue_type: str  # "error", "warning", "info"
    message: str
    expected_value: Optional[Any] = None
    actual_value: Optional[Any] = None


class ValidationSummary(BaseModel):
    is_valid: bool = True
    overall_confidence: float = 1.0
    errors_count: int = 0
    warnings_count: int = 0
    pdf_item_count: int = 0
    extracted_item_count: int = 0
    items_ready_for_excel: int = 0
    currency: str = "EUR"
    currency_symbol: str = "€"
    price_validation_passed: int = 0
    part_number_validation_passed: int = 0
    total_items_validated: int = 0
    issues: list[ValidationIssue] = Field(default_factory=list)


class ExtractionResponse(BaseModel):
    success: bool
    quotation: QuotationData
    validation: ValidationSummary
    processing_time_seconds: float = 0.0
    extracted_page_count: int = 1
    extraction_source: str = "pdf_text_ai"


class ColumnMappingConfig(BaseModel):
    line_number_column: Optional[str] = "A"
    part_number_column: Optional[str] = "B"
    description_column: Optional[str] = "C"
    quantity_column: Optional[str] = "D"
    unit_column: Optional[str] = None
    unit_price_column: Optional[str] = "F"
    discount_column: Optional[str] = None
    total_price_column: Optional[str] = None
    currency_column: Optional[str] = None
    start_row: int = 6
    quote_number_cell: Optional[str] = None
    customer_cell: Optional[str] = None
    date_cell: Optional[str] = None
    currency_cell: Optional[str] = None


class TemplateAnalysisResponse(BaseModel):
    template_id: str
    sheet_names: list[str]
    detected_headers: dict[str, str]
    suggested_mapping: ColumnMappingConfig
    preview_rows: list[list[Any]] = Field(default_factory=list)


class GenerateExcelRequest(BaseModel):
    quotation: QuotationData
    template_id: str
    pricing_mode: str = "quoted_price"
    margin_percent: float = 0.0
    supplier_discount_percent: Optional[float] = None
    custom_mapping: Optional[ColumnMappingConfig] = None


class GenerateExcelResponse(BaseModel):
    success: bool
    download_url: str
    filename: str
    file_id: str
    total_items_written: int
    grand_total_written: float

